### ISSUE TITLE
Add option to alternative MNI templates for output spaces

### ISSUE THREAD

**Issue by feilong** [poldracklab/fmriprep#487]
From discussions of another issue (https://github.com/poldracklab/fmriprep/issues/337#issuecomment-297139529), it would be nice to have alternatives to `MNI152NLin2009cAsym` as output spaces to increase backward compatibility with other softwares, datasets, and atlases.  For example, currently widely used `MNI152NLin6Asym`.

I'll try to modify `fmriprep/workflows/anatomical.py` and add `output_spaces` as an input parameter to `init_anat_preproc_wf`, and change the spatial normalization code accordingly.  Does that sound good? @chrisfilo

**Comment by effigies** [comment]
If I can make a suggestion, what might be nice would be to add a `--template-space={MNI152NLin2009cAsym,MNI152NLin6Asym,<etc>}` and change the MNI option in `--output-space` to `template`. This keeps `--output-space` options to a reasonably short list of manageable strings.

Unless we think people are potentially going to want their data in many template spaces?

**Comment by chrisfilo** [comment]
@feilong this change would require modifications  of the RobustMNINormalization Interface in niworkflows: https://github.com/poldracklab/niworkflows/blob/master/niworkflows/anat/mni.py#L35. The target atlas would also have to be included in the data downloader https://github.com/poldracklab/niworkflows/blob/master/niworkflows/data/getters.py#L19 (Although @oesteban might know if it is already included - if not if you can email it to me I will add it to our OSF repository).

@effigies Your proposal does reflect the fact that only one volumetric template would be allowed (which should be fine for 99% of users). I would however insist on keeping the full names of the template used in output file names (instead of "template") for provenance related reasons..

**Comment by effigies** [comment]
Yeah. Keeping the full name in filenames makes sense. Just from a UI perspective, not having to write out the full template unless changing from the default would be nice.

**Comment by effigies** [comment]
Actually, speaking of this, should creating structurals in MNI space depend on `--output-space`? Currently `t1_2_mni` and the `mni_*` nodes run no matter what.

**Comment by feilong** [comment]
@effigies I think so, and I guess it might be good to add template space names or abbreviations to those node names?

**Comment by oesteban** [comment]
@feilong do you refer to this atlas -> http://nist.mni.mcgill.ca/?p=858 ? It is not currently included in niworkflows. As @chrisfilo said, you'd need to add it.

**Comment by chrisfilo** [comment]
MNI152NLin6Asym is an asymmetrical version of this atlas - we would have to grab it from SPM or FSL.

**Comment by chrisfilo** [comment]
@feilong I'm trying to prioritize features - would you use/benefit from addition of `MNI152NLin6Asym`?

**Comment by feilong** [comment]
@chrisfilo Yes, I want my data resampled to both `MNI152NLin6Asym` and `MNI152NLin2009cAsym` if possible.

**Comment by effigies** [comment]
@chrisfilo is it worth updating `--template` to take multiple template specifiers?

**Comment by chrisfilo** [comment]
Interesting. So you would like outouts in both `MNI152NLin6Asym` and `MNI152NLin2009cAsym`? Despite it would take double the space?

We could either iterate over T1w coregistration targets or always use `MNI152NLin2009cAsym` and in the final stage apply one more non-linear warp from `MNI152NLin2009cAsym` to `MNI152NLin6Asym`. Such warp should not be hard to estimate.

**Comment by feilong** [comment]
I need some compatibility at this moment, but eventually `MNI152NLin2009cAsym` would be the choice.
I guess coregister to the target template directly would be more more accurate, though it would take a little longer.

**Comment by dangom** [comment]
Hi guys, would you have any updates on this issue? I was checking out fmriprep and was a bit frustrated that the template is fixed. For resting-state studies, I'd like to use a dual regression approach against maps generated in MNI 6th gen 2mm. I don't know the consequences of doing this if my maps are in MNI 2009.

Besides that,  ICA AROMA hyperplane parameters were developed for MNI 6th 2mm, as seen in the code [here](https://github.com/maartenmennes/ICA-AROMA/blob/master/ICA_AROMA_functions.py#L348) and [here](https://github.com/maartenmennes/ICA-AROMA/blob/master/ICA_AROMA_functions.py#L489) and in the Manual. Does this mean that, when the `--usa-aroma` flag is on, registration to MNI 6th is actually performed but then discarded? I reckon that the technique may also work in other templates, but it'd be nice to document that.

Finally, would there be a way of quickly hacking fmriprep into using MNI 6th 2mm by overwriting the MNI 2009 file? Any ideas if things would break?

**Comment by chrisfilo** [comment]
> Hi guys, would you have any updates on this issue? I was checking out fmriprep and was a bit frustrated that the template is fixed. For resting-state studies, I'd like to use a dual regression approach against maps generated in MNI 6th gen 2mm. I don't know the consequences of doing this if my maps are in MNI 2009.

Considering usual level of smoothness I would not expect a large difference, but we have not tested this explicitly. In principle we have nothing against adding an option to support other target templates (as you probably noticed the command line flag have been designed to accommodate that feature in the future). This could be something that you can help us to add by submitting a pull request.

> Besides that, ICA AROMA hyperplane parameters were developed for MNI 6th 2mm, as seen in the code here and here and in the Manual. Does this mean that, when the --usa-aroma flag is on, registration to MNI 6th is actually performed but then discarded? I reckon that the technique may also work in other templates, but it'd be nice to document that.

No - as indicated by the file naming convention of the outputs, transformations to 2009c template are used in the AROMA pipeline. I would be surprised if using MNI 6th would make a significant difference in this context.

> Finally, would there be a way of quickly hacking fmriprep into using MNI 6th 2mm by overwriting the MNI 2009 file? Any ideas if things would break?

It would break my heart :) We would much more appreciate if you could work towards implementing this as part of the core package extending the `--template` flag. We can provide guidance in the process.

> (by the way, thanks a lot for this project)
Thank you for the kind words and the feedback. You might also read this thread on neurostars

- https://neurostars.org/t/bids-fmriprep-transforms-to-3-5-x-3-75-x-3-75-mni-space/1391
- https://neurostars.org/t/fixing-inconsistent-output-spaces-across-studies-in-fmriprep-1-0-6/1309

**Comment by dangom** [comment]
Thanks for the quick and through answer. =)
What would be the necessary changes to get this to work? 

> this change would require modifications of the RobustMNINormalization Interface in niworkflows: https://github.com/poldracklab/niworkflows/blob/master/niworkflows/anat/mni.py#L35. The target atlas would also have to be included in the data downloader https://github.com/poldracklab/niworkflows/blob/master/niworkflows/data/getters.py#L19

Does the above still hold true?
I won't have time to tackle an implementation within the next few weeks, though.
